### PR TITLE
WPCOM SSH: Allow user to update an existing SSH key

### PR DESCRIPTION
**Pull request by kozer**
#### Proposed Changes

Right now, when we delete an custom SSH key, the key is detached from all the sites.
So when the user adds a new one, it's needed to go and reassign the ssh key to each site again.

Let the user update the ssh key, solves this problem.

Below are some screenshots of this new functionality.

image

image

##### Test 1

1. Make sure that you apply this patch to your sandbox: D91797-code
2. Sandbox public api.
3. Start calypso dev environment.
4. Navigate to `/me/security/ssh-key`
5. Add an ssh key.
6. Make sure that next to `Remove SSH Key` there is another button, named:`Update SSH Key`
7. Click that button
8. Make sure that a modal displaying `Are you sure you want to update this SSH key? It will affect all attached sites.` appears.
9. Press the OK button.
10. Make sure that a modal displaying the old ssh key and a textarea in order to enter a new one
   appears.
11. Enter a new SSH key and click `Update SSH Key`
12. Ensure that the button changes it's state.
13. After the new key is saved, ensure that the modal is auto closed and a banner indicating that
    the key was saved appears in the top right corner.

###### Test 2

1. Make sure that you apply this patch to your sandbox: D91797-code
2. Sandbox public api.
3. Start calypso dev environment.
4. Navigate to `/me/security/ssh-key`
5. Add an ssh key.
6. Make sure that next to `Remove SSH Key` there is another button, named:`Update SSH Key`
7. Click that button
8. Make sure that a modal displaying `Are you sure you want to update this SSH key? It will affect all attached sites.` appears.
9. Press the OK button.
10. Make sure that a modal displaying the old ssh key and a textarea in order to enter a new one
   appears.
11. Enter a new SSH key and click `Update SSH Key`
12. Ensure that the button changes it's state.
14. Close the modal before the ssh key is saved.
15. Ensure that the `Remove SSH Key button` is disabled.
16. Ensure that the `Update SSH Key` button can be pressed and you can view the modal.
17. Ensure that when the modal reopens, if the update is not yet completed, the textareas and the
    button are in the same state as when the modal was originally closed.

###### Test 3

1. Make sure that you apply this patch to your sandbox: D91797-code
2. Sandbox public api.
3. Start calypso dev environment.
4. Navigate to `/me/security/ssh-key`
5. Add an ssh key.
4. Switch your site to one that has SSH hosting enabled.
5. Navigate to hosting configuration  `/hosting-config/<your domain>` view
6. Make sure to attach the newly added SSH key to the site.
7. Click on your profile icon in the top right corner.
8. Click on `Security -> SSH Key`
9. Update the ssh key as described in `Test 1`
10. Click on `My Sites` on the top left corner
11. Click on `Settings -> Hosting Configuration`
12. Ensure that where previously it was attached the previous SSH key, now there is a loading
    indicator.
13. Wait till the loading is completed.
14. Ensure that the attached SSH key is not the previous one, but the updated SSH key instead

#### Pre-merge Checklist

- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?

Related to https://github.com/Automattic/dotcom-forge/issues/1135

### CODE REVIEW

**Comment by zaguiini** [comment]
I haven't tested it in conjunction with the diff. It looks good, and I think that using a modal instead of modifying the existing page makes more sense, because this way the user would be sure that just clicking the "Update SSH key" button won't change anything and they can cancel at any point.

How I imagine this working:

1. User opens the `/me/security/ssh-key` page;
2. User clicks on the "Update SSH key" button;
3. Modal opens with SSH key form, showing which key they are updating;
4. When clicking the submit button inside the modal, the prompt asking "are you sure about that?" pops up;
5. Once clicking yes, the submit button inside the modal shows the loading state;
6. The modal closes and the SSH keys are re-fetched, updating the SSH keys list.

Does that make sense?

Regarding MFA: we're already handling that on previous steps (e.g. when browsing this page/security settings), so I don't think you'll need to implement anything here.

**Comment by vykes-mac** [comment]
@zaguiini  @kozer I'm not sure if there was any discussion on this but I'm wondering why we giving users ability to update keys. The examples that I've seen is that a user can only remove an exisitng key and a new one if the want to change it or it expire, Eg. Github, DigitalOcean. I think ssh-keygen doesn't allow update of keys either only adding and removing (_could be wrong_) . Just wondering what benefit we get from updating as oppose to adding a new one which is what most services do, it seems.

**Comment by kozer** [comment]
@vykes-mac , yes you are right on that. The only benefit I see here is if we, behind the scenes, also make sure that we reassign the new ssh key to the sites that have assigned the old one. 
If not, it's actually like removing and adding a new one all in one step.
cc: @danielbachhuber

**Comment by kozer** [comment]
@zaguiini thanks for your suggestion. The only downside to that is that it will make the mobile experience poorer, but as we have already modals on that page, I think this is ok. 
Another downside to this is that we will provide two different experiences of doing the same thing: Adding an SSH key.
I'll go with the route you suggested though, as it will make things more clear, on what is editing and what's not.
However, I don't like the "prompt on modal" approach. Maybe prompt the user before opening the modal make things more clear?

**Comment by zaguiini** [comment]
I see @kozer. I think @vykes-mac is right, is not really common to see updates to SSH keys. I think removing and then re-adding, as two different steps for the user, is the way to go. We could even highlight this if necessary.

**Comment by kozer** [comment]
@zaguiini So maybe it's better to close the ticket for now? Or leave this open if we need to revisit it?

**Comment by danielbachhuber** [comment]
> The examples that I've seen is that a user can only remove an exisitng key and a new one if the want to change it or it expire, Eg. Github, DigitalOcean. I think ssh-keygen doesn't allow update of keys either only adding and removing (_could be wrong_) . Just wondering what benefit we get from updating as oppose to adding a new one which is what most services do, it seems.

@vykes-mac Great observation!

This is something I considered when I created the issue. The key difference between GitHub and WordPress.com is:
1. On GitHub, if you remove your SSH key, adding it again restores SSH access to all of your repositories.
2. On WordPress.com, removing your SSH key also removes it from all attached sites, which means you'll need to reattach it to each site after adding it again.

For this reason, I'd like to continue to implement the original issue.

> How I imagine this working:
> 
> 1. User opens the `/me/security/ssh-key` page;
> 2. User clicks on the "Update SSH key" button;
> 3. Modal opens with SSH key form, showing which key they are updating;
> 4. When clicking the submit button inside the modal, the prompt asking "are you sure about that?" pops up;
> 5. Once clicking yes, the submit button inside the modal shows the loading state;
> 6. The modal closes and the SSH keys are re-fetched, updating the SSH keys list.
> 
> Does that make sense?

@zaguiini I'm fine with this proposed approach. I don't think it's likely for someone to update their SSH key from a mobile device, so fine to switch to a modal.

**Comment by zaguiini** [comment]
>2. On WordPress.com, removing your SSH key also removes it from all attached sites, which means you'll need to reattach it to each site after adding it again.
> 
>For this reason, I'd like to continue to implement the original issue.

Just to clarify: the reason why you want to allow updating an SSH key is so one can set a new SSH key without detaching it [the named key] from the existing sites?

**Comment by danielbachhuber** [comment]
> the reason why you want to allow updating an SSH key is so one can set a new SSH key without detaching it [the named key] from the existing sites?

@zaguiini Yep!

**Comment by zaguiini** [comment]
> > the reason why you want to allow updating an SSH key is so one can set a new SSH key without detaching it [the named key] from the existing sites?
> 
> @zaguiini Yep!

Then one suggestion is to include some copy saying that updating the key means it won't detach from the sites that particular key is being used.

**Comment by kozer** [comment]
info-popover-update
@danielbachhuber @zaguiini , what do you think about that?

Also. Do we need to put the freezing strings label here right?

**Comment by danielbachhuber** [comment]
@kozer Eh, I don't love that popover. Let's simply incorporate @zaguiini's text suggestion into the existing body of text.

> Also. Do we need to put the freezing strings label here right?

Not in this case. It's not a frequently used UI, so it's fine to display the English strings for a while until the translations are complete.

**Comment by danielbachhuber** [comment]
@kozer Can you make sure to include screenshots of these changes in the description?

**Comment by zaguiini** [comment]
I'd like for us to display the update copy only when there's an existing key.

Also, the copy looks good:

>Updating an SSH key, it won't detach it from the sites that particular key is being used.

But I think it'd flow better as something like:

>Updating an SSH key won't detach it from the sites that particular key is being used.

I'd also move the copy to the last line, like that:

image

I still have mixed feelings about it, though. It looks like a lot of text that I'm not particularly happy about... How about moving the copy to the update form modal?

On another note, I'd appreciate more top spacing here:

image

Let's also remove the period:

image

**Comment by zaguiini** [comment]
When the textarea is in an invalid state, the modal stretches out:

image

One thing I noticed when updating the key is that for a second the modal gets in an intermediary state and bounces before closing. That happens because the existing key gets updated in the component before closing the modal. How about creating a copy of the value through a prop (i.e. `initialValue`) so that when the keys are prefetched, that value doesn't change?

**Comment by danielbachhuber** [comment]
> I still have mixed feelings about it, though. It looks like a lot of text that I'm not particularly happy about... How about moving the copy to the update form modal?

💯

**Comment by kozer** [comment]
> When the textarea is in an invalid state, the modal stretches out:
> 
> image
> 
> One thing I noticed when updating the key is that for a second the modal gets in an intermediary state and bounces before closing. That happens because the existing key gets updated in the component before closing the modal. How about creating a copy of the value through a prop (i.e. `initialValue`) so that when the keys are prefetched, that value doesn't change?

ssh-modal
What about that?

I have also some comments/questions:

> When the textarea is in an invalid state, the modal stretches out:
I couldn't find a good way to tackle this. Ohe thing I thought was to setup modal's width to be the same initially as when this text appears. But this makes the modal very wide (almost 1000px) which it wont work on smaller screens. Any ideas?

> One thing I noticed when updating the key is that for a second the modal gets in an intermediary state and bounces before closing. That happens because the existing key gets updated in the component before closing the modal. How about creating a copy of the value through a prop (i.e. initialValue) so that when the keys are prefetched, that value doesn't change?

Can you elaborate a bit more here? It would be beneficial if you showed me a video on that.

UPDATE: I 've found the problem in the modal. Scratch that. 

UPDATE2: It seems I can't remove the dot. If I do that I get a lint error: `[@wordpress/i18n-no-placeholders-only] Translatable strings should not contain nothing but placeholders (eslint)`

Any way to do so?

cc: @zaguiini

**Comment by zaguiini** [comment]
We're almost there, @kozer!

A couple of things I don't think are ideal:

1. There's no need to display the whole old public key there. How about showing the fingerprint and the key name? The modal gets too tall and even overflows at some point, making the update button disappear from the scrolling area.
2. I feel the update key copy could be a little more prominent, like within a box with a yellow background or something, indicating the information that is pretty important.

>It seems I can't remove the dot. If I do that I get a lint error:  [...] Any way to do so?

Yes, you'll need to pass the translated string as a prop instead of the string you want to translate. So, instead of:

```tsx
updateText="Update SSH Key"
```

and

```tsx
{ sprintf(
  // translators: "sshText" is the text of the ssh save button.
  __( '%(sshText)s.' ),
  {
    sshText: updateText || 'Save SSH Key',
  }
) }
```

It should be:

```tsx
updateText={ __( 'Update SSH Key' ) }
```

and

```tsx
{ updateText ?? __( 'Save SSH Key' ) }
```

In all places (`AddSSHKeyForm` and `UpdateSSHKeyForm`). That will fix it 👍 That error happens because you're trying to translate a string that only contained placeholders, and that doesn't make sense from an i18n perspective.

**Comment by kozer** [comment]
> We're almost there, @kozer!
> 
> A couple of things I don't think are ideal:
> 
> 1. There's no need to display the whole old public key there. How about showing the fingerprint and the key name? The modal gets too tall and even overflows at some point, making the update button disappear from the scrolling area.
> 2. I feel the update key copy could be a little more prominent, like within a box with a yellow background or something, indicating the information that is pretty important.
> 
> > It seems I can't remove the dot. If I do that I get a lint error:  [...] Any way to do so?
> 
> Yes, you'll need to pass the translated string as a prop instead of the string you want to translate. So, instead of:
> 
> ```tsx
> updateText="Update SSH Key"
> ```
> 
> and
> 
> ```tsx
> { sprintf(
>   // translators: "sshText" is the text of the ssh save button.
>   __( '%(sshText)s.' ),
>   {
>     sshText: updateText || 'Save SSH Key',
>   }
> ) }
> ```
> 
> It should be:
> 
> ```tsx
> updateText={ __( 'Update SSH Key' ) }
> ```
> 
> and
> 
> ```tsx
> { updateText ?? __( 'Save SSH Key' ) }
> ```
> 
> In all places (`AddSSHKeyForm` and `UpdateSSHKeyForm`). That will fix it +1 That error happens because you're trying to translate a string that only contained placeholders, and that doesn't make sense from an i18n perspective.

What do you think of that @zaguiini ?

modal-new

I tried to reuse things we already have for displaying the update key copy

**Comment by danielbachhuber** [comment]
@kozer For showing the old SSH key, let's reuse https://github.com/Automattic/wp-calypso/pull/69905

Also, as an alternative to the notice (which looks like an error to me), we could include a title and description for the modal:

> ## Update SSH Key
>
> Update your SSH key to change the SSH key used across all attached sites.

**Comment by danielbachhuber** [comment]
> For showing the old SSH key, let's reuse #69905

@kozer This is now available if you want to merge trunk into your feature branch.

**Comment by kozer** [comment]
@danielbachhuber 
What do you think of that?
modal-new

**Comment by danielbachhuber** [comment]
@kozer Better! I'd like to use a title and description for the modal too, please.

**Comment by kozer** [comment]
@danielbachhuber 
This is the modal updated. Title and description added. However I think that the "warning" alongside with the title are descriptive enough.
If you have any suggestions for the description / title please let me know.
modal-new-2

**Comment by kozer** [comment]
This is modal for now, after the discussions we had on this thread: p1668440441120489/1668439009.416639-slack-C7YPUHBB2

modal-new-4

**Comment by zaguiini** [comment]
Looking very good and tests nicely 👍 

Non-blocking suggestion: remove the line break between the two sentences inside the modal. Also, is there a way to limit how many lines the text area expands to? Do we need to show the whole key? If we do, how about making it slightly wider so when the key is pasted, less vertical space is taken?

Feel free to merge even without my explicit approval, this is looking good enough already! Way to go!

File: client/me/security-ssh-key/security-ssh-key.tsx
Comment (danielbachhuber) [inline client/me/security-ssh-key/security-ssh-key.tsx]: Can we use CSS objects, to follow patterns established elsewhere?

File: client/me/security-ssh-key/security-ssh-key.tsx
Comment (danielbachhuber) [inline client/me/security-ssh-key/security-ssh-key.tsx]:
```suggestion
			dispatch( successNotice( __( 'SSH key updated for account.' ), noticeOptions ) );
```

File: client/me/security-ssh-key/security-ssh-key.tsx
Comment (danielbachhuber) [inline client/me/security-ssh-key/security-ssh-key.tsx:137]: Can you create pull requests to record these events with Tracks?

File: client/me/security-ssh-key/manage-ssh-keys.tsx
Comment (danielbachhuber) [inline client/me/security-ssh-key/manage-ssh-keys.tsx]: Instead of having two modals, which is a little bit of an odd experience, let's simply add a 'Cancel' button to the modal with the form:

image

One modal would be much better.

File: client/me/security-ssh-key/manage-ssh-keys.tsx
Comment (danielbachhuber) [inline client/me/security-ssh-key/manage-ssh-keys.tsx:62]: This button shows up as red for me...

image

We should probably make 'Remove SSH key' the primary action, and 'Update SSH key' the secondary.

File: client/me/security-ssh-key/security-ssh-key.tsx
Comment (danielbachhuber) [inline client/me/security-ssh-key/security-ssh-key.tsx:228]: For the SSH modal description, let's use:
> Replace your current SSH key with a new SSH key to use the new SSH key with all attached sites.

File: client/me/security-ssh-key/update-ssh-key-form.tsx
Comment (danielbachhuber) [inline client/me/security-ssh-key/update-ssh-key-form.tsx]:
```suggestion
				<FormLabel>{ __( 'Current SSH Public Key' ) }</FormLabel>
```

File: client/me/security-ssh-key/update-ssh-key-form.tsx
Comment (danielbachhuber) [inline client/me/security-ssh-key/update-ssh-key-form.tsx]:
```suggestion
					<FormLabel htmlFor={ PUBLIC_SSH_KEY_INPUT_ID }>{ __( 'New SSH Public Key' ) }</FormLabel>
```

File: client/me/security-ssh-key/update-ssh-key-form.tsx
Comment (danielbachhuber) [inline client/me/security-ssh-key/update-ssh-key-form.tsx:53]: It seems like there's a lot of duplication between this and `<AddSSHKeyForm>`.

Is there any way we could consolidate both to a more generic `<SSHKeyForm>`?

**Review by danielbachhuber — COMMENTED** [review]
Solid progress, @kozer ! Left a few comments to work through.

File: client/me/security-ssh-key/update-ssh-key-form.tsx
Comment (kozer) [reply client/me/security-ssh-key/update-ssh-key-form.tsx:53]: They have many things in common, however, they have different structures. I believe in keeping it KISS instead of trying a more general component with strange props to keep it DRY.
If we found that there is a need to reuse the same logic to other places as well, I agree that maybe it worth the effort to put more thought into that.

File: client/me/security-ssh-key/security-ssh-key.tsx
Comment (kozer) [reply client/me/security-ssh-key/security-ssh-key.tsx:137]: Here is the pr for that

Update success:
https://github.com/Automattic/tracks-events-registration/pull/1175

Update failure
https://github.com/Automattic/tracks-events-registration/pull/1176

File: client/me/security-ssh-key/manage-ssh-keys.tsx
Comment (sejas) [inline client/me/security-ssh-key/manage-ssh-keys.tsx]: Since `keyBeingUpdated` is used only to calculate if the buttons should be disabled, maybe we could pass disabled (`!! keyBeingDeleted || !! keyBeingUpdated`) as a prop to `SSHKey` component, and the same variable can be used in both buttons. Line 62 and Line 71.

**Review by sejas — APPROVED** [review]
@kozer  Congrats !! It's looking great !!
I've added a tiny optimization totally optional.

Tested with the API D91797-code and both are working nicely ! Well done !! 👏

**Review by danielbachhuber — APPROVED** [review]
Nicely done on this, @kozer ! I really like the end result.

I made a few changes directly to your PR:
* [Use named endpoint for PUT request](https://github.com/Automattic/wp-calypso/pull/69826/commits/355aa2bdf0f99ffb5fc1baf0022624dc164306f8)
* [Disable the 'Update SSH key' button if the key is being deleted](https://github.com/Automattic/wp-calypso/pull/69826/commits/7ee118a6d139bacf647a3acbc61f58b1bcb9235d)
* [Use logical properties for RTL support](https://github.com/Automattic/wp-calypso/pull/69826/commits/27d3edc9b306f7cd1814f6770f4660dc8384f396)
* [Disable the 'Cancel' button when the SSH key is being updated](https://github.com/Automattic/wp-calypso/pull/69826/commits/6fbb99fcfda382b41f7a32786e3cf5960a972a07)

I think this is ready to ship after you land D91797-code

**Comment by kozer** [comment]
> Nicely done on this, @kozer ! I really like the end result.
> 
> I made a few changes directly to your PR:
> 
> * [Use named endpoint for PUT request](https://github.com/Automattic/wp-calypso/pull/69826/commits/355aa2bdf0f99ffb5fc1baf0022624dc164306f8)
> * [Disable the 'Update SSH key' button if the key is being deleted](https://github.com/Automattic/wp-calypso/pull/69826/commits/7ee118a6d139bacf647a3acbc61f58b1bcb9235d)
> * [Use logical properties for RTL support](https://github.com/Automattic/wp-calypso/pull/69826/commits/27d3edc9b306f7cd1814f6770f4660dc8384f396)
> * [Disable the 'Cancel' button when the SSH key is being updated](https://github.com/Automattic/wp-calypso/pull/69826/commits/6fbb99fcfda382b41f7a32786e3cf5960a972a07)
> 
> I think this is ready to ship after you land D91797-code

Thanks @danielbachhuber for your effort!

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7625922

Thank you @kozer for including a screenshot in the description! This is really helpful for our translators.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7625922).

**Pull request merged**
